extensions/amp-story/1.0/amp-story-tooltip.js

```diff
@@ -0,0 +1,401 @@
+      // desktop buttons.
+      this.closeTooltip_();
+    });
+
+    return this.shadowRoot_;
+  }
+
+  /**
+   * Hides the tooltip layer.
+   * @private
+   */
+  closeTooltip_() {
+    this.storeService_.dispatch(Action.TOGGLE_TOOLTIP, null);
+    // Hide active tooltip when page switch is triggered by keyboard or
+    // desktop buttons.
+    this.onTooltipStateUpdate_(false);
+  }
+
+  /**
+   * Reacts to store updates related to tooltip active status.
+   * @param {boolean} tooltipIsActive
+   * @private
+   */
+  onTooltipStateUpdate_(tooltipIsActive) {
+    this.resources_.mutateElement(dev().assertElement(this.tooltipOverlayEl_),
+        () => {
+          this.tooltipOverlayEl_
+              .classList.toggle('i-amphtml-hidden', !tooltipIsActive);
+        });
+  }
+
+  /**
+   * Reacts to desktop state updates and hides navigation buttons since we
+   * already have in the desktop UI.
+   * @param {!UIType} uiState
+   * @private
+   */
+  onUIStateUpdate_(uiState) {
+    this.resources_.mutateElement(dev().assertElement(this.tooltipOverlayEl_),
+        () => {
+          uiState === UIType.DESKTOP ?
+            this.tooltipOverlayEl_.setAttribute('desktop', '') :
+            this.tooltipOverlayEl_.removeAttribute('desktop');
+        });
+  }
+
+  /**
+   * Builds tooltip and attaches it depending on the clicked element content and
+   * position.
+   * @param {!Element} clickedEl
+   * @private
+   */
+  attachTooltipToEl_(clickedEl) {
+    const {href} = parseUrlDeprecated(clickedEl.getAttribute('href'));
+
+    const domainName = getSourceOriginForElement(clickedEl, href);
+    const tooltipText = clickedEl.getAttribute('tooltipText') || domainName;
+    this.appendTextToTooltip_(tooltipText);
+
+    const iconAttr = clickedEl.getAttribute('tooltipIcon');
+    const iconSrc = iconAttr ? parseUrlDeprecated(iconAttr).href : '';
+    this.appendIconToTooltip_(iconSrc);
+
+    this.positionTooltip_(clickedEl);
+  }
+
+  /**
+   * Checks for existing text content and modifies it or appends a new text node
+   * if building tooltip for the first time.
+   * @param {string} tooltipText
+   * @private
+   */
+  appendTextToTooltip_(tooltipText) {
+    const existingTooltipText = childElement(
+        dev().assertElement(this.tooltip_),
+        el => el.classList.contains('i-amphtml-tooltip-text'));
+
+    if (existingTooltipText) {
+      existingTooltipText.textContent = tooltipText;
+    } else {
+      const html = htmlFor(this.win_.document);
+      const tooltipTextEl =
+        html`<p class="i-amphtml-tooltip-text" ref="text"></p>`;
+      tooltipTextEl.textContent = tooltipText;
+      this.tooltip_.insertBefore(tooltipTextEl, this.tooltip_.firstChild);
+    }
+  }
+
+  /**
+   * Checks for existing icon source and modifies it or appends a new tooltip
+   * icon node if building tooltip for the first time.
+   * @param {string} iconSrc
+   * @private
+   */
+  appendIconToTooltip_(iconSrc) {
```

Comment:
It feels like the code would be a lot easier if the tooltip was built with the rest of the template, in the `build` method.
This method would then just set or remove the `src` attribute, and that's it.

Same comment for the `appendTextToTooltip_` method